Enhancements and Integrations for Kanto Service

frankplus opened the pull request:
This pull request introduces a series of updates and new integrations for the Kanto service. Key highlights include the setup of a system ability for Kanto, which is added to `bundle.json` for integration. The `kanto_service` has been configured within `kanto.cfg` to ensure that services start at system boot.

We've integrated the paho-mqtt library into the gn build system as a submodule, enhancing message handling capabilities. A `BUILD.gn` file has been created for the libpaho_mqtt_cpp library, and `VersionInfo.h`, typically generated by cmake, has been added manually.

Furthermore, we've introduced a basic application to the Kanto service, inspired by a sample from the Bosch IoT Kanto example applications. This sample demonstrates a simple console application that retrieves gateway device data via the Kanto connector and sends telemetry data for temperature readings from a simulated Air-Conditioning device.

Related-issue: https://gitlab.eclipse.org/eclipse/oniro-core/ohos-planning/-/issues/44

Commit:
add sa_profile for kanto

Add system ability profile for kanto and add it to bundle.json

Signed-off-by: Francesco Pham <francesco.pham@huawei.com>

Commit:
start kanto_service at startup

- add kanto_service to kanto.cfg in order to start services at system startup.
- Rename kanto service for suite-connector to "kanto_suite_connector"

Signed-off-by: Francesco Pham <francesco.pham@huawei.com>

Commit:
add basic kanto service application that uses paho mqtt

add basic application to kanto service taking from this sample: https://github.com/bosch-io/kanto-example-applications/tree/1-cpp-telemetry/temperature-sensor
This sample shows simple console application that gets the gateway device data from the Kanto connector and sends telemetry data for the temperature readings of a simulated Air-Conditioning device.

Signed-off-by: Francesco Pham <francesco.pham@huawei.com>

Stefan-Schmidt commented on services/kanto/include/kanto_hilog.h:
```diff
@@ -0,0 +1,88 @@
+    KANTO_MODULE_BUTT,
+};
+
+static constexpr unsigned int BASE_KANTO_DOMAIN_ID = 0xD001C04;
```
I am curious how this ID is selected. And how it avoid conflicts with other ID's that end up in the same system, but we do not know about.

frankplus replied:
thanks for pointing it out since the the base id is the same of another service ability. Fixed it now.
As to how it is selected, I think it's just random. Interesting to notice with a `grep` that different parts of the system have the same base domain id:
```
//base/request/request/frameworks/js/napi/include/upload/upload_hilog_wrapper.h:33:static constexpr unsigned int BASE_UPLOAD_DOMAIN_ID = 0xD001C00;
//foundation/distributeddatamgr/pasteboard/utils/native/include/pasteboard_hilog.h:35:static constexpr unsigned int BASE_PASTEBOARD_DOMAIN_ID = 0xD001C00;
``` 
I don't know how the logger is affected with conflicting ID

Stefan-Schmidt replied:
Interesting, still good to have a unique ID for Kanto.

Stefan-Schmidt commented on services/kanto/include/kanto_service.h:
```diff
@@ -0,0 +1,25 @@
+#ifndef KANTO_SERVICE_H_
```
Some of your files have a copyright statement, some are missing it.

frankplus replied:
added licenses to all source files now

Stefan-Schmidt commented on profile/8100.json:
```diff
@@ -0,0 +1,12 @@
+    "process": "kanto_service",
+    "systemability": [
+        {
+            "name": 8100,
```
Same question for the ID here. Is there some registry where these are coordinated?

frankplus replied:
the id must be defined in `foundation/systemabilitymgr/samgr/interfaces/innerkits/samgr_proxy/include/system_ability_definition.h` I will create a patch that adds this ID, and specify it in the README

Stefan-Schmidt replied:
Thanks

Stefan-Schmidt commented on services/kanto/src/kanto_service.cpp:
```diff
@@ -3,6 +3,8 @@
 #include <unistd.h>
 #include <system_error>
+#include "mqtt/client.h"
+#include "json.hpp"
```
Where does the json.hpp file come from? Gitlab web view does refuse to load it as it is to big with 24K lines. Origin and license would be interesting for the this. I assume its a one file json parser.

Stefan-Schmidt commented on services/kanto/src/kanto_service.cpp:
```diff
@@ -26,7 +28,77 @@ void KantoService::OnStop()
+        deviceIdInNamespaceNotation.replace(namespaceSeparatorIndex, 1, "/");
+        std::string telemetryTopic = "t/" + tenantId + "/" + airConditionerDeviceId;
+
+        nlohmann::json telemetryMessagePayload = {
```
With this hint I assume we are talking about hits file: https://github.com/nlohmann/json/blob/develop/single_include/nlohmann/json.hpp ?

We should note in the license description MIT next to Apache 2.0 now.

frankplus replied:
that is correct. Added MIT license to LICENSES now

Commit:
LICENSE: add MIT license for json.hpp file

 json.hpp has MIT license therefore adding the MIT to LICENSE

Signed-off-by: Francesco Pham <francesco.pham@huawei.com>

Commit:
add kanto service ability

Add simple service ability which is doing nothing but logging "starting kanto!"

Signed-off-by: Francesco Pham <francesco.pham@huawei.com>

Commit:
Add license headers

add Apache 2.0 license headers 
adjust license year 
add SoB to patch file

Signed-off-by: Francesco Pham <francesco.pham@huawei.com>

Commit:
add paho-mqtt as submodule, add BUILD.gn for paho library

- integrate paho-mqtt library to gn build adding it as a submodule
- create BUILD.gn adding libpaho_mqtt_cpp library.
- Add VersionInfo.h which is usually generated by cmake 

Signed-off-by: Francesco Pham <francesco.pham@huawei.com>
